+nosnoc/+time_freezing/cls_inelastic_multicontact.m

```diff
@@ -0,0 +1,293 @@
+    else
+        v_tangent  = [];
+    end
+    
+    % parameter for auxiliary dynamics
+    % TODO add to options
+    a_n = 100;
+    %% Time-freezing reformulation
+    % unconstrained dynamics with clock state
+    f_ode = [cls_model.v;...
+        cls_model.invM*cls_model.f_v;
+        f_quad;
+        1];
+
+    % Auxiliary dynamics
+    % where to use invM, in every auxiliary dynamics or only at the end
+    % TODO put this in problem opts
+    if true
+        inv_M_aux = eye(dims.n_q);
+        inv_M_ext = blkdiag(zeros(dims.n_q),cls_model.invM,zeros(1+dims.n_quad));
+    else
+        inv_M_aux = cls_model.invM;
+        inv_M_ext = eye(dims.n_x+1);
+    end
+    f_aux_pos = []; % matrix with all auxiliary tangent dynamics
+    f_aux_neg = [];
+    % time freezing dynamics
+    if opts.stabilizing_q_dynamics
+        f_q_dynamics = -opts.kappa_stabilizing_q_dynamics*cls_model.J_normal*diag(cls_model.f_c);
+    else
+        f_q_dynamics = zeros(dims.n_q,dims.n_contacts);
+    end
+    f_aux_normal = [f_q_dynamics;inv_M_aux*cls_model.J_normal*a_n;zeros(1+dims.n_quad,1)];
+
+    if opts.time_freezing_nonsmooth_switching_fun
+        heaviside_model.c = [max_smooth_fun(cls_model.f_c,v_normal,0);v_tangent];    
+    else
+        if dims.n_dim_contact == 2
+            heaviside_model.c = [cls_model.f_c;v_normal;v_tangent'];
+        else
+            heaviside_model.c = [cls_model.f_c;v_normal;v_tangent_norms-eps_t];
+        end
+    end
+
+    for ii = 1:dims.n_contacts
+        if cls_model.friction_exists && cls_model.mu(ii)>0
+            % auxiliary tangent;
+            if dims.n_dim_contact == 2
+                v_tangent_ii = cls_model.J_tangent(:,ii)'*cls_model.v;
+                f_aux_pos_ii = [f_q_dynamics(:,ii) ;inv_M_aux*(cls_model.J_normal(:,ii)-cls_model.J_tangent(:,ii)*(cls_model.mu(ii)))*a_n;zeros(1+dims.n_quad,1)]; % for v>0
+                f_aux_neg_ii = [f_q_dynamics(:,ii) ;inv_M_aux*(cls_model.J_normal(:,ii)+cls_model.J_tangent(:,ii)*(cls_model.mu(ii)))*a_n;zeros(1+dims.n_quad,1)]; % for v<0
+            else
+                v_tangent_ii = v_tangent(:,ii);
+                f_aux_pos_ii = [f_q_dynamics(:,ii);inv_M_aux*(cls_model.J_normal(:,ii)*a_n-cls_model.J_tangent(:,ii*2-1:ii*2)*cls_model.mu(ii)*a_n*v_tangent_ii/norm(v_tangent_ii+1e-12));zeros(1+dims.n_quad,1)]; % for v>0
+                f_aux_neg_ii = [f_q_dynamics(:,ii);inv_M_aux*(cls_model.J_normal(:,ii)*a_n+cls_model.J_tangent(:,ii*2-1:ii*2)*cls_model.mu(ii)*a_n*v_tangent_ii/norm(v_tangent_ii+1e-12));zeros(1+dims.n_quad,1)]; % for v>0
+            end
+            f_aux_pos = [f_aux_pos,f_aux_pos_ii];
+            f_aux_neg = [f_aux_neg,f_aux_neg_ii];
+        end
+    end
+    if cls_model.friction_exists
+        f_aux = [f_aux_pos,f_aux_neg];
+        dims.n_aux = 2*dims.n_contacts;
+    else
+        f_aux = f_aux_normal;
+        dims.n_aux = dims.n_contacts;
+    end
+
+    % Build logical functions because its easier than hand picking logic.
```

Comment:
a lot of shared core. again the dilemma do we extract them into a function. i have no strong opinion.

Reply:
Yea extracted into a few functions. The interfacing is not the cleanest but that is ok probably to trade for one source of truth.